packages/frontend/src/models/CanonicalBridge.ts

```diff
@@ -0,0 +1,341 @@
+import { Signer, ethers, BigNumber, BigNumberish, utils } from 'ethers'
+import {
+  ArbitrumInbox__factory,
+  ArbitrumInbox,
+  L1ArbitrumDaiGateway__factory,
+  L1ArbitrumDaiGateway,
+  L1OptimismDaiTokenBridge__factory,
+  L1OptimismTokenBridge__factory,
+  L1OptimismTokenBridge,
+  L1PolygonPlasmaBridgeDepositManager__factory,
+  L1PolygonPlasmaBridgeDepositManager,
+  L1PolygonPosRootChainManager__factory,
+  L1PolygonPosRootChainManager,
+  L1XDaiForeignOmniBridge__factory,
+  L1XDaiForeignOmniBridge,
+  L1XDaiPoaBridge__factory,
+  L1XDaiPoaBridge,
+  L1XDaiWETHOmnibridgeRouter__factory,
+  L1XDaiWETHOmnibridgeRouter,
+} from '@hop-protocol/core/contracts'
+import { ChainSlug, CanonicalToken } from '@hop-protocol/sdk'
+import Base, { L1Factory } from '@hop-protocol/sdk/dist/src/Base'
+import { metadata } from '@hop-protocol/sdk/dist/src/config'
+import { TokenSymbol } from '@hop-protocol/sdk/dist/src/constants'
+import { Chain } from '@hop-protocol/sdk/dist/src/models'
+import Token from '@hop-protocol/sdk/dist/src/models/Token'
+import TokenClass from '@hop-protocol/sdk/dist/src/Token'
+import { TChain, TToken } from '@hop-protocol/sdk/dist/src/types'
+import { Bridge } from 'arb-ts'
+import { JsonRpcSigner } from '@ethersproject/providers'
+
+export type L1CanonicalBridge =
+  | ArbitrumInbox
+  | L1OptimismTokenBridge
+  | L1PolygonPlasmaBridgeDepositManager
+  | L1PolygonPosRootChainManager
+  | L1XDaiForeignOmniBridge
+  | L1XDaiPoaBridge
+  | L1XDaiWETHOmnibridgeRouter
+
+class CanonicalBridge extends Base {
+  public chain: Chain
+  public tokenSymbol: TokenSymbol
+  address: string
+  l2TokenAddress: string
+
+  constructor(network: string, signer: Signer, token: TToken, chain: TChain) {
+    super(network, signer)
+    if (!token) {
+      throw new Error('token symbol is required')
+    }
+    this.getSignerOrProvider(Chain.Ethereum, signer).then(s => {
+      this.signer = s
+    })
+    this.chain = this.toChainModel(chain)
+    this.tokenSymbol = this.toTokenModel(token).symbol
+    this.address = this.getL1CanonicalBridgeAddress(this.tokenSymbol, this.chain)
+    this.l2TokenAddress = this.getL2CanonicalTokenAddress(this.tokenSymbol, this.chain)
+  }
+
+  public connect(signer: Signer) {
+    return new CanonicalBridge(this.network, signer, this.tokenSymbol, this.chain)
+  }
+
+  public getL1CanonicalAllowance(): Promise<BigNumber> {
+    const l1CanonicalToken = this.getL1Token().connect(this.signer)
```

Comment:
suggestion: move `.connect(this.signer)` be to `getL1Token` method to avoid having to manually connect every time `getL1Token` is called